crates/biome_js_analyze/src/lint/nursery/use_explicit_type.rs

```diff
@@ -740,3 +719,105 @@ fn is_type_assertion(syntax: &SyntaxNode<JsLanguage>) -> bool {
         }
     })
 }
+
+fn handle_any_function(func: &AnyJsFunction) -> Option<TextRange> {
+    if func.return_type_annotation().is_some() {
+        return None;
+    }
+
+    if is_direct_const_assertion_in_arrow_functions(func) {
+        return None;
+    }
+
+    if is_iife(func) {
+        return None;
+    }
+
+    if is_function_used_in_argument_or_array(func) {
+        return None;
+    }
+
+    if is_arrow_func(func) && (can_inline_function(func) || is_function_inside_typed_return(func)) {
+        return None;
+    }
+
+    if is_higher_order_function(func) {
+        return None;
+    }
+
+    if is_typed_function_expressions(func) {
+        return None;
+    }
+
+    let func_range = func.syntax().text_trimmed_range();
+    if let Ok(Some(AnyJsBinding::JsIdentifierBinding(id))) = func.id() {
+        return Some(TextRange::new(
+            func_range.start(),
+            id.syntax().text_trimmed_range().end(),
+        ));
+    }
+
+    Some(func_range)
+}
+
+/// Checks if a variable declarator needs to have an explicit type.
+fn handle_variable_declarator(
+    declarator: &JsVariableDeclarator,
+    model: &SemanticModel,
+) -> Option<TextRange> {
+    let variable_declaration = declarator
+        .parent::<JsVariableDeclaratorList>()?
+        .parent::<JsVariableDeclaration>()?;
+    let is_top_level = declarator
+        .syntax()
+        .ancestors()
+        .find_map(JsVariableStatement::cast)
+        .is_some_and(|statement| {
+            statement
+                .syntax()
+                .parent()
+                .is_some_and(|parent| JsModuleItemList::can_cast(parent.kind()))
+        })
+        || variable_declaration
+            .parent::<JsVariableDeclarationClause>()
+            .is_some();
+
+    let identifier = declarator.id().ok()?;
+    let identifier = identifier.as_any_js_binding()?.as_js_identifier_binding()?;
+    let is_exported = model.is_exported(identifier);
+
+    if !is_top_level || !is_exported {
+        return None;
+    }
+
+    let initializer_expression = declarator
+        .initializer()
+        .and_then(|init| init.expression().ok())
+        .map(|expr| expr.omit_parentheses());
+
+    let is_const = variable_declaration.is_const();
+
+    let ty = declarator
+        .variable_annotation()
+        .and_then(|ty| ty.as_ts_type_annotation().cloned())
+        .and_then(|ty| ty.ty().ok());
+
+    if let Some(ty) = ty {
+        if let Some(initializer_expression) = initializer_expression {
+            if (is_const && ty.is_non_null_literal_type())
+                || (!is_const
+                    && ty.is_primitive_type()
+                    && !matches!(
+                        initializer_expression,
+                        AnyJsExpression::AnyJsLiteralExpression(
+                            AnyJsLiteralExpression::JsNullLiteralExpression(_)
+                        )
+                    ))
+            {
+                return None;
+            }
+        }
+    }
```

Comment:
What the purpose of this?

Also, we should certainly ignore code with trivially inferrable types?

Reply:
This code is the same logic we use for those types that can be inferred, as you requested.